ambry-replication/src/main/java/com.github.ambry.replication/CloudToStoreReplicationManager.java

```diff
@@ -0,0 +1,333 @@
+    replicationMetrics.addMetricsForRemoteReplicaInfo(remoteReplicaInfo);
+    List<RemoteReplicaInfo> remoteReplicaInfos = Collections.singletonList(remoteReplicaInfo);
+    PartitionInfo partitionInfo = new PartitionInfo(remoteReplicaInfos, partitionId, store, localReplicaId);
+    partitionToPartitionInfo.put(partitionId, partitionInfo);
+    mountPathToPartitionInfos.computeIfAbsent(localReplicaId.getMountPath(), key -> ConcurrentHashMap.newKeySet())
+        .add(partitionInfo);
+    logger.info("Cloud Partition {} added to {}", partitionName, dataNodeId);
+
+    // Reload replication token if exist.
+    List<RemoteReplicaInfo.ReplicaTokenInfo> tokenInfos = persistor.retrieve(localReplicaId.getMountPath());
+    if (tokenInfos.size() != 0) {
+      for (RemoteReplicaInfo remoteInfo : remoteReplicaInfos) {
+        boolean tokenReloaded = false;
+        for (RemoteReplicaInfo.ReplicaTokenInfo tokenInfo : tokenInfos) {
+          if (isTokenForRemoteReplicaInfo(remoteInfo, tokenInfo)) {
+            logger.info("Read token for partition {} remote host {} port {} token {}", partitionId,
+                tokenInfo.getHostname(), tokenInfo.getPort(), tokenInfo.getReplicaToken());
+            tokenReloaded = true;
+            remoteInfo.initializeTokens(tokenInfo.getReplicaToken());
+            remoteInfo.setTotalBytesReadFromLocalStore(tokenInfo.getTotalBytesReadFromLocalStore());
+            break;
+          }
+        }
+        if (!tokenReloaded) {
+          // This may happen on clusterMap update: replica removed or added.
+          // Or error on token persist/retrieve.
+          logger.warn("Token not found or reload failed. remoteReplicaInfo: {} tokenInfos: {}", remoteInfo, tokenInfos);
+        }
+      }
+    }
+    // Add remoteReplicaInfos to {@link ReplicaThread}.
+    addRemoteReplicaInfoToReplicaThread(remoteReplicaInfos, true);
+    if (replicationConfig.replicationTrackPerPartitionLagFromRemote) {
+      replicationMetrics.addLagMetricForPartition(partitionId);
+    }
+  }
+
+  /**
+   * Remove a replica of given partition and its {@link RemoteReplicaInfo}s from the backup list.
+   * @param partitionName the partition of the replica to removed.
+   * @throws ReplicationException if replicas initialization failed.
+   */
+  private void removeCloudReplica(String partitionName) throws ReplicationException {
+    if (!localPartitionNameToPartition.containsKey(partitionName)) {
+      logger.error("Got partition standby notification for partition {} that is not present on the node",
+          partitionName);
+      return;
+    }
+    PartitionId partitionId = localPartitionNameToPartition.get(partitionName);
+    PartitionInfo partitionInfo = partitionToPartitionInfo.remove(partitionId);
+    if (partitionInfo == null) {
+      logger.error("Partition {} not exist when remove from {}. ", partitionId, dataNodeId);
+      throw new ReplicationException("Partition not found");
```

Comment:
Combine full error message into exception and remove log.

Reply:
fixed.